ggshield/oauth.py

```diff
@@ -0,0 +1,273 @@
+import os
+import urllib.parse as urlparse
+import webbrowser
+from base64 import urlsafe_b64encode
+from datetime import datetime
+from hashlib import sha256
+from http.server import BaseHTTPRequestHandler, HTTPServer
+from secrets import token_urlsafe
+from typing import Any, Dict, Optional, Type
+
+import click
+import requests
+from oauthlib.oauth2 import OAuth2Error, WebApplicationClient
+
+from .client import retrieve_client
+from .config import AccountConfig, Config
+
+
+CLIENT_ID = "ggshield_oauth"
+SCOPE = "scan"
+REDIRECT_PORT = 1234
+
+
+class OAuthClient:
+    def __init__(self, config: Config, instance: str) -> None:
+        self.config = config
+        self.instance = instance
+        self._oauth_client = WebApplicationClient(CLIENT_ID)
+        self._access_token: Optional[str] = None
+
+        self.server: Optional[HTTPServer] = None
+        self._handler_wrapper = RequestHandlerWrapper(oauth_client=self)
+        self._generate_pkce_pair()
+
+    def oauth_process(self) -> None:
+        """
+        Handle the whole oauth process which includes
+        - opening the user's webbrowser to GitGuardian login page
+        - open a server and wait for the callback processing
+        """
+        # enable redirection to http://localhost
+        os.environ["OAUTHLIB_INSECURE_TRANSPORT"] = str(True)
+
+        self._token_name = "ggshield token " + datetime.today().strftime("%Y-%m-%d")
+
+        self._prepare_server()
+        self._redirect_to_login()
+        self._wait_for_code()
+        click.echo(f"Created Personal Access Token {self._token_name}")
+
+    def process_callback(self, callback_url: str) -> None:
+        """
+        This function runs within the request handler do_GET method.
+        It takes the url of the callback request as argument and does
+        - Extract the authorization code
+        - Exchange the code against an access token with GitGuardian's api
+        - Validate the new token against GitGuardian's api
+        - Save the token in configuration
+        Any error during this process will raise a OAuthError
+        """
+        authorization_code = self._get_code(callback_url)
+        self._claim_token(authorization_code)
+        token_data = self._validate_access_token()
+        self._save_token(token_data)
+
+    def _generate_pkce_pair(self) -> None:
+        """
+        Generate a code verifier (random string) and its sha encoded version to be used
+        for the pkce checking process
+        """
+        self.code_verifier = token_urlsafe(96)[:128]
```

Comment:
:question: what's the point of `[:128]`? it's a fixed number of bytes, it should be the same character length for the accepted characters, no?

Reply:
In fact I'll revert to the `oauthlib`'s `create_code_verifier` method
https://github.dev/oauthlib/oauthlib/blob/b123283ba3d41acb3e787fdf68bd5907972b4bad/oauthlib/oauth2/rfc6749/clients/base.py#L468